### ISSUE TITLE
Custom Fields not being registered

### ISSUE THREAD

**Issue by erossdev** [strapi/strapi#16875]
### Required System information

- Node.js version: v18.15.0
- NPM version: 9.5.0
- Strapi version: 4.10.7
- Database: SQLite
- Operating system: Mac OS, M1 chip
- Is your project Javascript or Typescript: Javascript

### Describe the bug

I created a new plugin with `npm run strapi generate` and added the most simple plugin i could but i still receive `Error: Could not find Custom Field: plugin::test.test`. I am able to add the custom field to a content-type but once I do, I begin to get that error

### Steps to reproduce the behavior

1. Create a new strapi project
2. `npm run strapi generate` a plugin
3. Define a component/Input (I just put a TextInput in there)
4. register a custom field on the server and on the client
5. Add the custom field to a content type on the admin panel

On the admin client
```javascript
app.customFields.register({
        name: "test",
        pluginId: "test", // the custom field is created by a color-picker plugin
        type: "string", // the color will be stored as a string
        intlLabel: {
            id: "test.test.label",
            defaultMessage: "Color",
        },
        intlDescription: {
            id: "test.test.description",
            defaultMessage: "Select any color",
        },
        icon: PluginIcon, // don't forget to create/import your icon component 
        components: {
            Input: async () => import(/* webpackChunkName: "input-component" */ "./components/Input"),
        },
        options: {
            // declare options here
        },
    });
```

On the admin server in register.js
```javascript
module.exports = ({ strapi }) => {
  strapi.customFields.register({
      name: "test",
      pluginId: "test", // the custom field is created by a color-picker plugin
      type: "string", // the color will be stored as a string
  });
};
```

Getting an error
```
Error: Could not find Custom Field: plugin::test.test
    at Object.get (/Users/esro/projects/wwo-raffle-website/strapi-backend/node_modules/@strapi/strapi/lib/core/registries/custom-fields.js:34:15)
    at convertCustomFieldType (/Users/esro/projects/wwo-raffle-website/strapi-backend/node_modules/@strapi/strapi/lib/utils/convert-custom-field-type.js:15:67)
    at Strapi.register (/Users/esro/projects/wwo-raffle-website/strapi-backend/node_modules/@strapi/strapi/lib/Strapi.js:408:5)
    at async Strapi.load (/Users/esro/projects/wwo-raffle-website/strapi-backend/node_modules/@strapi/strapi/lib/Strapi.js:491:5)
    at async workerProcess (/Users/esro/projects/wwo-raffle-website/strapi-backend/node_modules/@strapi/strapi/lib/commands/actions/develop/action.js:110:26)

```

### Expected behavior

The plugin should be registered and usable

### Code snippets

If applicable, add code samples to help explain your problem.

**Comment by erossdev** [comment]
Oh wow...I just realized that when registering your plugin on the server it is `plugin` not `pluginId`

WRONG WAY
```javascript
module.exports = ({ strapi }) => {
  strapi.customFields.register({
      name: "test",
      pluginId: "test", //  <-- this should be plugin: "test"
      type: "string", 
  });
};
```

CORRECT WAY
```javascript
module.exports = ({ strapi }) => {
  strapi.customFields.register({
      name: "test",
      plugin: "test",
      type: "string", 
  });
};
```

**Comment by esiao** [comment]
I would like to reopen this issue, as it might help others. I ran into the same problem but due to Typescript type checking.

In the `register.ts` file

```
import { Strapi } from '@strapi/strapi';

export default ({ strapi }: { strapi: Strapi }) => {
  strapi.customFields.register({
    name: 'link',
    plugin: 'link-custom-field',
    type: 'json',
  });
};
```

This throws an error because plugin doesn't exist in `CustomFieldServerOptions`.

I thought I fixed it by renaming plugin to pluginId, and the ran into the issue reported here. My solution so far is the following:
```
export default ({ strapi }: { strapi: Strapi }) => {
  strapi.customFields.register({
    name: 'link',
    pluginId: 'link-custom-field',
    // @ts-expect-error
    plugin: 'link-custom-field',
    type: 'json',
  });
};
```

I believe the `CustomFieldServerOptions` has to be updated so it works client/server side.

**Comment by xenobytezero** [comment]
Can confirm I hit the exact same issue when updating to 4.13.6. Having to any cast my `strapi.customFields.register` calls

**Comment by Wasenshi123** [comment]
> Can confirm I hit the exact same issue when updating to 4.13.6. Having to any cast my `strapi.customFields.register` calls

+1

**Comment by BorysShulyak** [comment]
I have the same issue

**Comment by xenobytezero** [comment]
Strapi 4.14 seems to fix this issue.

**Comment by AlexisSniffer** [comment]
For me it works
version: 4.15.5

**plugins/product-variants/admin/src/index.tsx**
```javascript
register(app: any) {
    app.customFields.register({
      name: 'variants',
      pluginId: 'product-variants',
      type: 'string',
      intlLabel: {
        id: getTrad('product-variants.label'),
        defaultMessage: 'Variants',
      },
      intlDescription: {
        id: getTrad('product-variants.description'),
        defaultMessage: 'Variants (Cartesian product)',
      },
      components: {
        Input: async () =>
          import(
            /* webpackChunkName: "input-component" */ './components/Variants/Input'
          ),
      },
      icon: PluginIcon,
      options: {},
    })
},
```

**plugins/product-variants/server/register.ts**
```javascript
export default ({ strapi }: { strapi: Strapi }) => {
  strapi.customFields.register({
    name: 'variants',
    plugin: 'product-variants',
    type: 'string',
    inputSize: {
      default: 4,
      isResizable: true,
    },
  })
}
```
config/plugins.ts
```javascript
// prettier-ignore
'product-variants': {
  enabled: true,
  resolve: './src/plugins/product-variants'
},
```

**Comment by derrickmehaffy** [comment]
Marking this as resolved in 4.15.5
